internal/getproviders/package_authentication.go

```diff
@@ -65,29 +92,192 @@ func NewArchiveChecksumAuthentication(wantSHA256Sum [sha256.Size]byte) PackageAu
+	}
+
+	// Success! But this doesn't result in any real authentication, only a
+	// lack of authentication errors, so we return a nil result.
+	return nil, nil
+}
+
+type signatureAuthentication struct {
+	Document  []byte
+	Signature []byte
+	Keys      []SigningKey
+}
+
+// NewSignatureAuthentication returns a PackageAuthentication implementation
+// that verifies the cryptographic signature for a package against a given key.
+func NewSignatureAuthentication(document, signature []byte, keys []SigningKey) PackageAuthentication {
+	return signatureAuthentication{
+		Document:  document,
+		Signature: signature,
+		Keys:      keys,
+	}
+}
+
+func (s signatureAuthentication) AuthenticatePackage(meta PackageMeta, location PackageLocation) (*PackageAuthenticationResult, error) {
+	if _, ok := location.(PackageLocalArchive); !ok {
+		// A source should not use this authentication type for non-archive
+		// locations.
+		return nil, fmt.Errorf("cannot check archive hash for non-archive location %s", location)
 	}
-	return nil
+
+	if _, ok := meta.Location.(PackageHTTPURL); !ok {
+		// A source should not use this authentication type for non-HTTP source
+		// locations.
+		return nil, fmt.Errorf("cannot check archive hash for non-HTTP location %s", meta.Location)
+	}
+
+	// Attempt to verify the signature using each of the keys returned by the
+	// registry. Note: currently the registry only returns one key, but this
+	// may change in the future. We must check each key in turn to find the
+	// matching signing entity before proceeding.
+	var signingKey *SigningKey
+	for _, key := range s.Keys {
+		keyring, err := openpgp.ReadArmoredKeyRing(strings.NewReader(key.ASCIIArmor))
+		if err != nil {
+			return nil, err
+		}
+
+		_, err = openpgp.CheckDetachedSignature(keyring, bytes.NewReader(s.Document), bytes.NewReader(s.Signature))
+
+		// If the signature issuer does not match the the key, keep trying the
+		// rest of the provided keys.
+		if err == openpgpErrors.ErrUnknownIssuer {
+			continue
+		}
+
+		// Any other signature error is terminal.
+		if err != nil {
+			return nil, err
+		}
+
+		signingKey = &key
+		break
+	}
+
+	// If none of the provided keys issued the signature, this package is
+	// unsigned. This is currently a terminal authentication error.
+	if signingKey == nil {
+		return nil, fmt.Errorf("Authentication signature from unknown issuer")
+	}
+
+	// Verify the signature using the HashiCorp public key. If this succeeds,
+	// this is an official provider.
+	hashicorpKeyring, err := openpgp.ReadArmoredKeyRing(strings.NewReader(HashicorpPublicKey))
+	if err != nil {
+		return nil, fmt.Errorf("Error creating HashiCorp Partners keyring: %s", err)
+	}
+	_, err = openpgp.CheckDetachedSignature(hashicorpKeyring, bytes.NewReader(s.Document), bytes.NewReader(s.Signature))
+	if err == nil {
+		return &PackageAuthenticationResult{Result: "HashiCorp provider"}, nil
```

Comment:
This might not make sense, depending on how we use this PackageAuthenticationResult elsewhere, but it might be a good idea to make the result a defined type instead of a string, something like:

`Result:`  ( `HashiCorpProvider` | `VerifiedProvider` | `CommunityProvider` )

Then we can type check on that and add a `.String()` function for user-facing output.

Reply:
I'm absolutely open to disagreement here :)

Reply:
This is also what I was thinking here: https://github.com/hashicorp/terraform/blob/registry-provider-tiers/plugin/discovery/tier.go

Reply:
Thanks, this makes sense! I'll revisit this type.

Reply:
I think 633e049 addresses this feedback. Thanks again!